### PR TITLE
Add abstractions for Juju's local config.

### PR DESCRIPTION
**Pull request by ericsnowcurrently**
Add Config, ControllerConfig, CloudConfig, and BootstrapConfig.

This PR is based on #2.

### CODE REVIEW

File: txjuju/config.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        namedtuple("CloudConfig",
+                   "name type endpoint auth_types credentials")):
+    """An encapsulation of the local config for a single cloud."""
+
+    def __new__(cls, name, type=None, endpoint=None,
+                auth_types=None, credentials=None):
+        """
+        @param name: The cloud's user-defined ID.
+        @param type: The provider type (defaults to the name).
+        @param endpoint: The endpoint to use, if needed.
```
Comment (blackboxsw) [inline txjuju/config.py]: can we add simple examples in the docstring of expected formats for these parameters.
For example, should endpoint begin w/ http://?
Shall we mention type should be one of lxd, maas, openstack etc?

File: txjuju/config.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
+# Copyright 2016 Canonical Limited.  All rights reserved.
+
+from collections import namedtuple
+
+
+class Config(object):
```
Comment (blackboxsw) [inline txjuju/config.py]: I thought we were missing Config class unit tests but it seems that test coverage comes in pull request 4 so I'll review that too.

File: txjuju/config.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        if self.auth_types is not None or self.credentials is not None:
+            raise NotImplementedError
+
+
+# TODO Implement CloudCredentialConfig as soon as needed.
+
+
+class BootstrapConfig(
+        namedtuple("BootstrapConfig", "default_series admin_secret")):
+    """An encapsulation of a bootstrap config."""
```
Comment (blackboxsw) [inline txjuju/config.py]: In landscape we end up setting bootstrap-timeout, and provisioner-harvest-mode: "destroyed" as well as setting logging-config. Is that worth surfacing these parameters in this class?

File: txjuju/config.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            cls, name, cloud, bootstrap)
+
+    def __init__(self, *args, **kwargs):
+        super(ControllerConfig, self).__init__(*args, **kwargs)
+        if not self.name:
+            raise ValueError("missing name")
+        if not self.cloud:
+            raise ValueError("missing cloud")
+        if not self.bootstrap:
+            raise ValueError("missing bootstrap")
```
Comment (blackboxsw) [inline txjuju/config.py]: not covered in unit tests.

File: txjuju/config.py (hunk 3 above)
Comment (ericsnowcurrently) [inline txjuju/config.py]: My initial focus is on fake-juju.  We can circle back on this when we update landscape to use txjuju.config.

File: txjuju/config.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        @param endpoint: The endpoint to use, if needed.
+        @param auth_types: The set of supported auth types, if needed.
+        @param credentials: The set of credentials configured
+            for this cloud, if any.
+        """
+        name = unicode(name) if name else None
+        if type is None:
+            type = name
+        type = unicode(type) if type else None
+        endpoint = unicode(endpoint) if endpoint else None
```
Comment (ericsnowcurrently) [inline txjuju/config.py]: There isn't any stable heuristic other than "what juju accepts".

**Comment by ericsnowcurrently** [comment]
@blackboxsw I've addressed all your review comments.  Thanks for the review (and approval). :)

**Pull request merged**